client.go

```diff
@@ -36,32 +43,158 @@ const (
 	return m.Name + "_" + m.UniqueKey
 }
 
-// DialV1Server is a convenience function for dialing gubernator instances
-func DialV1Server(server string, tls *tls.Config) (V1Client, error) {
-	if len(server) == 0 {
-		return nil, errors.New("server is empty; must provide a server")
+type ClientOptions struct {
+	// Users can provide their own http client with TLS config if needed
+	Client *http.Client
+	// The address of endpoint in the format `<scheme>://<host>:<port>`
+	Endpoint string
+}
+
+type client struct {
+	*duh.Client
+	prop propagation.TraceContext
+	opts ClientOptions
+}
+
+// NewClient creates a new instance of the Gubernator user client
+func NewClient(opts ClientOptions) (Client, error) {
+	setter.SetDefault(&opts.Client, DefaultHTTPClient)
+
+	if len(opts.Endpoint) == 0 {
+		return nil, errors.New("opts.Endpoint is empty; must provide an address")
+	}
+
+	return &client{
+		Client: &duh.Client{
+			Client: opts.Client,
+		},
+		opts: opts,
+	}, nil
+}
+
+func NewPeerClient(opts ClientOptions) PeerClient {
+	return &client{
+		Client: &duh.Client{
+			Client: opts.Client,
+		},
+		opts: opts,
+	}
+}
+
+func (c *client) CheckRateLimits(ctx context.Context, req *CheckRateLimitsRequest, resp *CheckRateLimitsResponse) error {
+	payload, err := proto.Marshal(req)
+	if err != nil {
+		return duh.NewClientError(fmt.Errorf("while marshaling request payload: %w", err), nil)
 	}
 
-	// Setup OpenTelemetry interceptor to propagate spans.
-	opts := []grpc.DialOption{
-		grpc.WithStatsHandler(otelgrpc.NewClientHandler()),
+	r, err := http.NewRequestWithContext(ctx, http.MethodPost,
+		fmt.Sprintf("%s%s", c.opts.Endpoint, RPCRateLimitCheck), bytes.NewReader(payload))
+	if err != nil {
+		return duh.NewClientError(err, nil)
 	}
-	if tls != nil {
-		opts = append(opts, grpc.WithTransportCredentials(credentials.NewTLS(tls)))
-	} else {
-		opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	r.Header.Set("Content-Type", duh.ContentTypeProtoBuf)
+	return c.Do(r, resp)
+}
+
+func (c *client) HealthCheck(ctx context.Context, resp *HealthCheckResponse) error {
+	payload, err := proto.Marshal(&HealthCheckRequest{})
+	if err != nil {
+		return duh.NewClientError(fmt.Errorf("while marshaling request payload: %w", err), nil)
 	}
 
-	conn, err := grpc.Dial(server, opts...)
+	r, err := http.NewRequestWithContext(ctx, http.MethodPost,
+		fmt.Sprintf("%s%s", c.opts.Endpoint, RPCHealthCheck), bytes.NewReader(payload))
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to dial server %s", server)
+		return duh.NewClientError(err, nil)
 	}
 
-	return NewV1Client(conn), nil
+	r.Header.Set("Content-Type", duh.ContentTypeProtoBuf)
+	return c.Do(r, resp)
+}
+
+func (c *client) Forward(ctx context.Context, req *ForwardRequest, resp *ForwardResponse) error {
+	payload, err := proto.Marshal(req)
+	if err != nil {
+		return duh.NewClientError(fmt.Errorf("while marshaling request payload: %w", err), nil)
```

Comment:
More a suggestion for `duh` - it might be helpful for this method to accept the error and wrap for you, like: `duh.NewClientError(err, "while marshaling request payload")`. Tiny tiny stylistic item for reducing code read time (and absolutely not required!)

Reply:
This. ☝🏻